File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/domain/entity/CollectCategory.java
package com.windea.demo.cloudcollect.core.domain.entity;

import com.windea.demo.cloudcollect.core.validation.annotation.UniqueCollectCategory;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.lang.NonNull;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.time.LocalDateTime;

/**
 * 收藏的分类。
 * <p>
 * 一个收藏只能带有一个分类。
 */
@Data
@NoArgsConstructor
@UniqueCollectCategory
@Entity
public class CollectCategory implements Serializable {
	private static final long serialVersionUID = -8860788073635562232L;

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	/**
	 * 名字。
	 */
	@NonNull
	@NotEmpty(message = "validation.CollectCategory.name.NotEmpty")
	@Size(min = 1, max = 32, message = "validation.CollectCategory.name.Size")
	@Column(nullable = false, length = 32)
	private String name = "";

	/**
	 * 概述。
	 */
	@NonNull
	@NotEmpty(message = "validation.CollectCategory.summary.NotEmpty")
	@Size(min = 1, max = 255, message = "validation.CollectCategory.summary.Size")
	@Column(nullable = false)
	private String summary = "";

	/**
	 * 所属用户。
	 */
	@NonNull
	@ManyToOne(cascade = CascadeType.MERGE, fetch = FetchType.EAGER, optional = false)
	private User user;

	/**
	 * 创建时间。
	 */
	@CreatedDate
	@Column
	private LocalDateTime createdTime;

	/**
	 * 最后更新时间。
	 */
	@LastModifiedDate
	@Column
	private LocalDateTime lastModifiedTime;

	public CollectCategory(String name, String summary, User user) {
		this.name = name;
		this.summary = summary;
		this.user = user;
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/domain/entity/CollectCategory.kt
package com.windea.demo.cloudcollect.core.domain.entity

import com.windea.demo.cloudcollect.core.validation.annotation.*
import com.windea.demo.cloudcollect.core.validation.group.*
import io.swagger.annotations.*
import org.springframework.data.annotation.*
import org.springframework.data.jpa.domain.support.*
import java.io.*
import java.time.*
import javax.persistence.*
import javax.persistence.Id
import javax.validation.constraints.*

@ApiModel("收藏的分类。一个收藏可以有多个分类。")
@Entity
@EntityListeners(AuditingEntityListener::class)
@UniqueCollectCategory(groups = [Create::class])
data class CollectCategory(
	@ApiModelProperty("编号。")
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	var id: Long? = null,
	
	@ApiModelProperty("名字。")
	@field:NotEmpty(message = "{validation.CollectCategory.name.NotEmpty}", groups = [Create::class, Modify::class])
	@field:Size(min = 1, max = 64, message = "{validation.CollectCategory.name.Size}", groups = [Create::class, Modify::class])
	@Column(nullable = false, length = 64)
	var name: String,
	
	@ApiModelProperty("概述。")
	@field:NotEmpty(message = "{validation.CollectCategory.summary.NotEmpty}", groups = [Create::class, Modify::class])
	@field:Size(min = 1, max = 255, message = "{validation.CollectCategory.summary.Size}", groups = [Create::class, Modify::class])
	@Column(nullable = false)
	var summary: String = "",
	
	@ApiModelProperty("所属用户。")
	@ManyToOne(cascade = [CascadeType.MERGE], fetch = FetchType.EAGER, optional = false)
	var user: User
) : Serializable {
	@ApiModelProperty("创建时间。")
	@Column
	@CreatedDate
	lateinit var createdTime: LocalDateTime
	
	@ApiModelProperty("最后更新时间。")
	@Column
	@LastModifiedDate
	lateinit var lastModifiedTime: LocalDateTime
	
	@ApiModelProperty("收藏数量。")
	@Transient
	var collectCount: Long = 0
	
	
	override fun equals(other: Any?) = other === this || (other is CollectCategory && other.id == id)
	
	override fun hashCode() = id.hashCode()
}
